src/runner/math-patches.js

```diff
@@ -0,0 +1,97 @@
+const DEC_128 = "decimal128";
+const BIG_DECIMAL = "bigdec";
+
+const createUnaryHandler = (substituteFns) => ({
+  apply(target, thisArg, argsList) {
+    const [arg] = argsList;
+    if (arg instanceof Decimal) {
+      return substituteFns[DEC_128](arg);
+    }
+
+    if (arg instanceof Big) {
+      return substituteFns[BIG_DECIMAL](arg);
+    }
+
+    return target.apply(thisArg, argsList);
+  },
+});
+
+const createNaryHandler = (decimalSubstituteFn) => ({
+  apply(target, thisArg, argsList) {
+    const decimals = argsList.filter((arg) => arg instanceof Decimal);
+
+    if (decimals.length < argsList.length) {
+      throw new TypeError(`mixed types not supported in Math.${mathMethod}`);
+    }
+
+    if (decimals.length === 0) {
+      return target.apply(thisArg, argsList);
+    }
+
+    return decimalSubstituteFn(...argsList);
+  },
+});
+
+// A few functions can take mixed values (max, min, pow)
+// The refiner can be used to apply more careful rules about mixing
+const createNaryMixedHandler = (substituteFns, refiner = () => true) => ({
```

Comment:
Haven't written a `refiner` but it would do things like check that for `pow` only the second arg is a `Number`, etc.